Adding router-server integration tests

cgtz opened the pull request:
Adding test cases and framework to test router operations with actual server (Issue #303):
- Framework for composing tests with chains of router operations
- Test interleaved/noninterleaved get, put, and del operations
- Test compatibility between coordinator-backed and non-blocking router
  operations
- Test large object support

**Reviewers:** @pnarayanan @xiahome 
**Time to review:** 30 min.

vgkholla commented on build.gradle:
```diff
@@ -127,6 +127,7 @@ project(':ambry-server') {
     dependencies {
         compile project(':ambry-clustermap'),
                 project(':ambry-coordinator'),
+                project(':ambry-router'),
```
think this can be unit test compile only - since the production code doesn't depend on router.

xiahome commented:
Some failures I saw on my mac. Would be good if this PR could be updated with more info on these.

com.github.ambry.server.RouterServerPlaintextTest > largeBlobTest FAILED
    java.lang.AssertionError: Get and Put blob content should match, operation: getBlob-nb-0 expected:<-85> but was:<0>
        at org.junit.Assert.fail(Assert.java:91)
        at org.junit.Assert.failNotEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:126)
        at org.junit.Assert.assertEquals(Assert.java:470)
        at com.github.ambry.server.RouterServerTestFramework.checkBlob(RouterServerTestFramework.java:142)
        at com.github.ambry.server.RouterServerTestFramework.access$200(RouterServerTestFramework.java:47)
        at com.github.ambry.server.RouterServerTestFramework$4.check(RouterServerTestFramework.java:217)
        at com.github.ambry.server.RouterServerTestFramework.checkFutures(RouterServerTestFramework.java:85)
        at com.github.ambry.server.RouterServerPlaintextTest.largeBlobTest(RouterServerPlaintextTest.java:112)

com.github.ambry.server.RouterServerSSLTest > interleavedOperationsTest FAILED
    java.lang.AssertionError: Blob should have been deleted in operation: getBlobInfo-deleted-nb-0
        at org.junit.Assert.fail(Assert.java:91)
        at com.github.ambry.server.RouterServerTestFramework$3.check(RouterServerTestFramework.java:191)
        at com.github.ambry.server.RouterServerTestFramework.checkFutures(RouterServerTestFramework.java:85)
        at com.github.ambry.server.RouterServerSSLTest.interleavedOperationsTest(RouterServerSSLTest.java:88)

cgtz commented:
I updated the PR to make the tests after deletions more stable.  I also rebased on top of the message format changes, so the tests should now work without additional changes.

nsivabalan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,140 @@
+  @BeforeClass
+  public static void initializeTests()
+      throws Exception {
+//    Logger.getRootLogger().setLevel(Level.ERROR);
```
why commented out

cgtz replied:
I left that in by mistake.  I was printing logs to aid with testing but I should just remove these lines.

nsivabalan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,481 @@
+     */
+    AWAIT_DELETION(false, false);
+
+    final boolean nonBlocking;
```
name this nonBlockingRouter. Even the cooridnator backed router is non blocking. So this is little mis-leading

cgtz replied:
I will fix this.

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+import org.junit.Assert;
+
+
+class RouterServerTestFramework {
```
I think some documentation would help for this class. Just a little about what the framework does, how it is used etc.
Documentation is good in general unless the method is trivial.

pnarayanan commented:
I think you can remove the bit about large object support requiring messageformat changes, as you have rebased it with those changes now.

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+    }
+  }
+
+  void checkOperationChains(List<OperationInfo> opInfos)
```
docs would be good.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,140 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import org.apache.log4j.Level;
```
This and the next shown not being used in my ide.

cgtz replied:
removed

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,140 @@
+import static com.github.ambry.server.RouterServerTestFramework.getRouterProperties;
+
+
+public class RouterServerPlaintextTest {
```
For all integration tests, I always have the question that what would be hardware requirement for running a number of real servers simultaneously. That also translates to another question that is what is the right configuration for the clustermap so everything should be kept under control.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,140 @@
+      throws Exception {
+    List<OperationInfo> opInfos = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      Queue<OperationType> opChain = new LinkedList<>();
```
This seems a pretty random operation sequence. Instead of looping for 100 times, my suggestion is to try some different sequences for completeness. I did some small test e.g., inserting one more `opChain.add(OperationType.GET_INFO_NB);` after line 71 it still passes, but this forms a different operation sequence and increases the coverage for test cases.

cgtz replied:
I will decrease the number of operation chains and add some more variation in the sequences themselves.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  @Test
+  public void nonInterleavedOperationsTest()
```
What would be the case that could not be covered by `interleavedOperationsTest()`? Or, what could be the error that only can be identified by this test?

cgtz replied:
This case is probably not that useful on its own but could help provide debugging insight if Interleaved fails but NonInterleaved passes.  However, it might be better to just leave this test case out by default.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  @Test
+  public void largeBlobTest()
```
This test takes a long time to finish on my mac, more than 30s.

cgtz replied:
Yes, this test does take a long time. Should I try decreasing the blob size to speed it up?

xiahome replied:
Yes, that might be better.

cgtz replied:
fixed, blob size is now `2 * chunk_size + 1` bytes

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+    double blobsPut = 0;
+    for (OperationInfo opInfo : opInfos) {
+      opInfo.latch.await();
+      synchronized (opInfo.futures) {
```
Why `synchronized` is necessary here?

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+      opInfo.latch.await();
+      synchronized (opInfo.futures) {
+        for (TestFuture future : opInfo.futures) {
+          future.check();
```
Suggest rename to `testFuture` just to be very specific. Since you have `future` as well for the router operation result. Correspondingly, ln. 385 make some change.

cgtz replied:
done

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+    }
+    double blobBalanceThreshold = 3.0 * Math.ceil(blobsPut / (double) clusterMap.getWritablePartitionIds().size());
+    for (Map.Entry<PartitionId, Integer> entry : partitionCount.entrySet()) {
+      Assert.assertTrue("Number of blobs is " + entry.getValue() + " on partition: " + entry.getKey()
```
Do we really have a control in load balancing among partitions? Would it be an error if the load is not that balanced?

cgtz replied:
This threshold is pretty conservative (3 times expected value).  Also, the partition chosen should be assigned randomly so ideally it should be close to balanced.

xiahome replied:
I see. So do we have a concrete meaning for a test failure here?

cgtz replied:
The test will fail if the number of blobs put on a specific partition is 3 times greater than the `(total_blobs_put / num_partitions_in_cluster)`

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+              operationInfo.data[readBytes++], buf.get());
+        }
+        getChannel.resolveOldestChunk(null);
+      } while (readBytes < operationInfo.data.length);
```
For any misbehavior if a router returns bytes less than expected, would this loop be running forever?

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+              try {
+                future.get();
+                Assert.fail("Blob should have been deleted in operation: " + getOperationName());
+              } catch (Exception ignored) {
```
Maybe we should assert the type of exception is what we expect.

cgtz replied:
Yes, that's a good idea.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+  }
+
+  private void startAwaitCreation(final OperationInfo opInfo) {
+    notificationSystem.awaitBlobCreations(opInfo.blobId);
```
The `awaitBlobCreations()` method has a specific timeout (which is 2 second for now), and this could potentially makes the test to run unfavorably long time. I do not see a good way to fix, but just raise a flag here.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+  }
+
+  private void continueChain(final OperationInfo opInfo) {
+    synchronized (opInfo.futures) {
```
Could you explain why `synchronized` is needed here? My understanding is that `continueChain` within the same operation chain will get executed sequentially, so there should be no race condition.

cgtz replied:
So, it's true that `continueChain` calls within a single operation chain are sequential.  However, consider the scenario where we are executing the last operation in a chain.  There is the slight possibility that this last operation can complete and call its callback before this callback has finished executing.  This means that the operation can be marked as completed (by flipping the latch) before the last operation's future was added to the list of futures.  This means that the check method might miss checking the last future in a chain.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+  /**
+   * Describes an operation chain and provides useful metadata and operation results for testing.
+   */
+  static class OperationInfo {
```
As discussed earlier today, this class is really not about info of an operation, but info of an operation chain. Please consider rename it to avoid confusion.

cgtz replied:
done

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+    final byte[] userMetadata;
+    final byte[] data;
+    final Queue<OperationType> opChain;
+    final List<TestFuture> futures = new ArrayList<>();
```
`futures` rename to `testFutures`.

cgtz replied:
done

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+    T get() {
+      try {
+        return this.future.get();
+      } catch (Exception e) {
```
@vgkholla has a good practice that whenever there is an unexpected exception, we simply throw without cooking it.

cgtz replied:
I agree that it is generally good practice.  However, here I wanted to provide the operation name (operation type, chain ID) in the assertion message for debugging purposes.  What do you think?

vgkholla replied:
rethrow the exception if you really want to do that.

vgkholla replied:
and are you sure you don't have a better way to provide this information (operation etc)?

cgtz replied:
I think this is the best way to provide this information, so I have changed this function to wrap the exception and rethrow it.  If you have a suggestion for a better approach, I'm pretty open.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,476 @@
+    @Override
+    public void onCompletion(T result, Exception exception) {
+      if (exception != null && !expectError) {
+        opInfo.latch.countDown();
```
It seems this part is never touched. What if falling within this block?

cgtz replied:
This block will complete the operation chain if an unexpected error occurs in an operation.

xiahome commented:
The `interleavedOperationsTest` takes at least 20 seconds in my mac and I have not seen a sign to stop. However it should finish because if I change loops to 2 or 10 it can be done in a reasonably short time. Please identify where is blocking and see how this could be improved.

xiahome commented:
Done with plain text test. SSL to be reviewed.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  @Test
+  public void interleavedOperationsTest()
```
Do we have a way to verify that ssl feature is really picked up? It could be possible that even all ssl parameters are configured, but they are not really being used.

xiahome commented:
Mostly LGTM, except I have some questions related to the comments I gave. Tomorrow I will come to you @cgtz  and discuss.

vgkholla commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,577 @@
+          void check() {
+            if (afterDelete) {
+              try {
+                future.get();
```
please change all `get()` to timed ones.

cgtz replied:
fixed

vgkholla commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,577 @@
+
+    @Override
+    public void onCompletion(T result, Exception exception) {
+      if (exception != null && !expectError) {
```
how does the caller know what exception happened via the callback?

cgtz replied:
The caller has to check the future to see what exception happened.

vgkholla replied:
That would defeat the purpose of having a callback right?

cgtz replied:
I was using the callback to either submit the next operation in the chain asynchronously or end the operation chain if an exception is encountered or if there are no operations left in the chain.  I wanted assertions to be done from the main test thread so that is why I don't do much with the exception in the callback.

vgkholla commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,577 @@
+      } catch (Exception e) {
+        e.printStackTrace();
+        Assert.fail("Exception occured in operation: " + getOperationName() + ", exception: " + e);
+        return null;
```
if you have an `fail()`, this will be an unreachable statement.

cgtz replied:
The exception is now rethrown, so the compiler no longer requires a return value past this point.

vgkholla commented:
Will go through this soon.

xiahome commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,577 @@
+    Map<PartitionId, Integer> partitionCount = new HashMap<>();
+    double blobsPut = 0;
+    for (OperationChain opChain : opChains) {
+      opChain.latch.await();
```
@vgkholla made a good point that it is always good to put a timeout in await(), so we do not risk to have the test run forever. Something like:
```
if(opChain.latch.await(!1000, TimeUnit.MILLISECONDS) {
  // create some exception here.
}
```

cgtz replied:
fixed

cgtz commented:
@xiahome @vgkholla addressed comments

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,612 @@
+    properties.setProperty("router.hostname", "localhost");
+    properties.setProperty("router.datacenter.name", routerDatacenter);
+    properties.setProperty("router.connection.checkout.timeout.ms", "5000");
+    properties.setProperty("router.request.timeout.ms", "20000");
```
did you need to override this? why would it require 20 seconds?

cgtz replied:
I had set a large timeout because multi-chunk operations were taking a long time.  I can make this smaller because I made the large blobs smaller to speed up the test.

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,612 @@
+    Map<PartitionId, Integer> partitionCount = new HashMap<>();
+    double blobsPut = 0;
+    for (OperationChain opChain : opChains) {
+      opChain.latch.await(AWAIT_TIMEOUT, TimeUnit.SECONDS);
```
don't you need to check the return value of `await()` here and perhaps fail the test?

cgtz replied:
Yes, I forgot that await does not throw an exception like timed `future.get()` calls.

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,612 @@
+    }
+    double numPartitions = clusterMap.getWritablePartitionIds().size();
+    if (opChains.size() > numPartitions) {
+      double blobBalanceThreshold = 3.0 * Math.ceil(blobsPut / numPartitions);
```
perhaps you can define the factor of `3.0` as a constant?

cgtz replied:
done

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,616 @@
+   */
+  private static void checkBlobInfo(BlobInfo blobInfo, OperationChain opChain, String operationName) {
+    Assert.assertNotNull("Null blobInfo for operation: " + operationName, blobInfo);
+    Assert.assertEquals("Blob size in info does not match expected for operation: " + operationName,
```
`RouterTestHelpers.java` has a method called `haveEquivalentFields()` for comparing `BlobProperties`, do you want to use that? (just fyi, we decided not to override `equals()` for `BlobProperties` itself due to the fact that it would break equality when comparison happens between `BlobProperties` and its derived class `SystemMetadata`).

cgtz replied:
Okay, I'll copy that method over and use it, unless there is a way to add the router test classes as a dependency.

pnarayanan replied:
Ah, these are in the server package. If you don't have the dependency now, I suggest leaving this as is.

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java:
```diff
@@ -0,0 +1,616 @@
+    }
+  }
+
+  private static String genLabel(String name, boolean nonBlocking, boolean afterDelete) {
```
javadoc with a bit of context on what this label is for and such?

cgtz replied:
will add

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  @Test
+  public void interleavedOperationsTest()
```
javadocs for non-trivial methods?

cgtz replied:
will add

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,157 @@
+          operations.add(OperationType.GET_INFO_NB);
+          break;
+      }
+      opChains.add(testFramework.startOperationChain(32 * 1024, i, operations));
```
I think this value should be defined as a constant (maybe even randomize it between runs?).

cgtz replied:
changed to a randomized value

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  @Test
+  public void nonInterleavedOperationsTest()
```
javadocs.

cgtz replied:
will add

pnarayanan commented:
@cgtz Nicely done. I just made a few more comments, looks good to me otherwise.

cgtz commented:
@pnarayanan addressed comments

pnarayanan commented:
LGTM. Merging.
